sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java

```diff
@@ -0,0 +1,195 @@
+      }
+      Event<T> event = events.get(index);
+      if (event.getType().equals(EventType.WATERMARK)) {
+        currentWatermark = ((WatermarkEvent<T>) event).getWatermark();
+      }
+      StepTransformResult.Builder result =
+          StepTransformResult.withHold(application, currentWatermark);
+      if (event.getType().equals(EventType.ELEMENT)) {
+        UncommittedBundle<T> bundle = context.createRootBundle(application.getOutput());
+        for (TimestampedValue<T> elem : ((ElementEvent<T>) event).getElements()) {
+          bundle.add(WindowedValue.timestampedValueInGlobalWindow(elem.getValue(),
+              elem.getTimestamp()));
+        }
+        result.addOutput(bundle);
+      }
+      if (event.getType().equals(EventType.PROCESSING_TIME)) {
+        ((TestClock) context.getClock())
+            .advance(((ProcessingTimeEvent<T>) event).getProcessingTimeAdvance());
+      }
+      index++;
+      checkState(inUse.compareAndSet(true, false),
+          "The InUse flag of a %s was changed while the source evaluator was executing. "
+              + "%s cannot be split or evaluated in parallel.",
+          TestStream.class.getSimpleName(),
+          TestStream.class.getSimpleName());
+      return result.build();
+    }
+  }
+
+  private static class TestClock implements Clock {
+    private final AtomicReference<Instant> currentTime =
+        new AtomicReference<>(BoundedWindow.TIMESTAMP_MIN_VALUE);
+
+    public void advance(Duration amount) {
+      Instant now = currentTime.get();
+      currentTime.compareAndSet(now, now.plus(amount));
+    }
+
+    @Override
+    public Instant now() {
+      return currentTime.get();
+    }
+  }
+
+  private static class TestClockSupplier implements Supplier<Clock> {
+    @Override
+    public Clock get() {
+      return new TestClock();
+    }
+  }
+
+  static class DirectTestStreamFactory implements PTransformOverrideFactory {
+    @Override
+    public <InputT extends PInput, OutputT extends POutput> PTransform<InputT, OutputT> override(
+        PTransform<InputT, OutputT> transform) {
+      if (transform instanceof TestStream) {
+        return (PTransform<InputT, OutputT>)
+            new DirectTestStream<OutputT>((TestStream<OutputT>) transform);
+      }
+      return transform;
+    }
+
+    private static class DirectTestStream<T> extends PTransform<PBegin, PCollection<T>> {
+      private final TestStream<T> original;
+
+      private DirectTestStream(TestStream transform) {
+        this.original = transform;
+      }
+
+      @Override
+      public PCollection<T> apply(PBegin input) {
+        setup(input.getPipeline());
+        return PCollection.<T>createPrimitiveOutputInternal(
+                input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED)
+            .setCoder(original.getValueCoder());
+      }
+
+      private void setup(Pipeline p) {
```

Comment:
Not entirely clear why this had to be broken out -- were you intending other places to call setup?

Seems like this does both some validation and some mutation. It seems odd that having a DirectTestStream mutates the clock supplier for the whole pipeline. What if you had both a DirectTestStream and other non-DirectTestStream sources?